## Задание 1

In [10]:
import pandas as pd

h_data = pd.read_csv('horses_stats.csv',na_values = '?', names = ['surgery', 'age','hospital_number', 'rectal_temperature', 'pulse', 'resp_rate', 'temp_of_extrem', 'peripheral_pulse', 'mucous_membranes', 'capil_refill_t', 'pain', 'peristalsis','abdominal_distension', 'nasogastric_tube', 'nasogastric_reflux', 'nasogastric_reflux_PH', 'rectal_examination_feces', 'abdomen', 'packed_cell_volume', 'total_protein', 'abdominocentesis_appearance', 'abdomcentesis_total_protein', 'outcome', 'surgical_lesion', 'type_of_lesion_1', 'type_of_lesion_2', 'type_of_lesion_3', 'cp_data'])


In [14]:
h_data.info()

i = 1
for col in h_data.columns:
    pct_missing = h_data[col].isnull().mean()
    print(f'{i} {col} - {pct_missing :.1%}')
    i += 1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery                      299 non-null    float64
 1   age                          300 non-null    int64  
 2   hospital_number              300 non-null    int64  
 3   rectal_temperature           240 non-null    float64
 4   pulse                        276 non-null    float64
 5   resp_rate                    242 non-null    float64
 6   temp_of_extrem               244 non-null    float64
 7   peripheral_pulse             231 non-null    float64
 8   mucous_membranes             253 non-null    float64
 9   capil_refill_t               268 non-null    float64
 10  pain                         245 non-null    float64
 11  peristalsis                  256 non-null    float64
 12  abdominal_distension         244 non-null    float64
 13  nasogastric_tube    

За пустое значение None был принят знак вопроса после просмотра загруженных данных, что поможет нам определить не репрезентативные столбцы для их исключения.

Далее рассмотрим процент пропущенных значений для более простого анализа исключаемых столбцов. Возьмем за критическое значение 30%, но с рассмотрением возможной пользы каждого столбца с значением пропусков, превышающих критическое.

Данные nosograstic..., abdomen, abdominocentesis_appearance, abdomcentesis_total_protein (14-18, 21-22) нерепрезентативны, выкидываем.

Первые 5 факторов важны для исследования и репрезентативны, оставляем (под вопросом только номер госпиталя).

6 спорный, как отмечается в описании, также нет 20% данных, выкидываем

7-9 - субъективные и коррелирующие, исключаем.

10 - ?

11-13 оставляем

14-18 выкидываем

19-20 ?

21-22 выкидываем

23 оставляем

из 25-27 можем понять что было в 24, дальнейшее исследование нужны ли

Итого оставляем 1-2, 4-5, 11-13, 19-20, 23, что составляет 10 столбцов.

In [15]:
main_factors = h_data[['surgery', 'age','rectal_temperature', 'pulse',  'pain', 'peristalsis','abdominal_distension','packed_cell_volume', 'total_protein','outcome']]

In [19]:
main_factors.describe()

surgery         age  rectal_temperature       pulse       pain  \
count  299.000000  300.000000          240.000000  276.000000  245.00000   
mean     1.397993    1.640000           38.167917   71.913043    2.95102   
std      0.490305    2.173972            0.732289   28.630557    1.30794   
min      1.000000    1.000000           35.400000   30.000000    1.00000   
25%      1.000000    1.000000           37.800000   48.000000    2.00000   
50%      1.000000    1.000000           38.200000   64.000000    3.00000   
75%      2.000000    1.000000           38.500000   88.000000    4.00000   
max      2.000000    9.000000           40.800000  184.000000    5.00000   

       peristalsis  abdominal_distension  packed_cell_volume  total_protein  \
count   256.000000            244.000000          271.000000     267.000000   
mean      2.917969              2.266393           46.295203      24.456929   
std       0.976744              1.065131           10.419335      27.475009   
min       1.000000              1.000000           23.000000       3.300000   
25%       3.000000              1.000000           38.000000       6.500000   
50%       3.000000              2.000000           45.000000       7.500000   
75%       4.000000              3.000000           52.000000      57.000000   
max       4.000000              4.000000           75.000000      89.000000   

          outcome  
count  299.000000  
mean     1.551839  
std      0.737187  
min      1.000000  
25%      1.000000  
50%      1.000000  
75%      2.000000  
max      3.000000

Судя по статистике, чаще лошади были прооперированы, большинство лошадей, попавших в госпиталь - молодые, младше 6 месяцев. 

Также в возрасте присутствует ошибочное значение = 9, тк принимаемые значения 1 и 2.

Средняя температура 38,1, что приближено к норме = 37,8. 

Средний пульс значительно выше нормальных показателей в покое - 30-40 ударов в минуту. Также присутствует значение 184, что нужно исследовать по другим параметрам, был ли у лошади болевой шок или же это выброс и ошибочное значение.

Срежний показатель боли показывает, что лошади в основном находятся в состоянии слабой боли.

Также в среднем лошади испытывают проблемы с животом, тк значение приближается к оценке их состояния, как болезненное.

В среднем у лошади присутствует слабое вздутие живота.

Наличие красных клеток в крови 46,3 приближается к верхнему предельному значению (50).

Значительное превышение значения протеина в органзиме - 24 к 6-7. Остается вопрос, это нормально и как раз с этим связаны болезни лошадей, или же есть ошибка в данных, слишком большое стандартное отклонение и максимальное значение.

В основном лошади после операции оставались живы, тк среднее значение последствий приближается к 1.

## Задание 2

Будем определять выбросы по 4 показателям - температура, пульс, наличие красных клеток, количество протеина в организме, а также разберемся с колонкой возраста.
Границы будем определять при помощи формулы:
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)

Вывод: 

1) Выбросы по столбцу температуры возникли в связи с тем, что интервал определения выбросов температуры достаточно узкий, около 2 градусов, в связи с чем в него не попало 11 значений температуры. Из данных в таблице ниже мы видим, что значения незначительно отличаются от пороговых значений как в одну, так и в другую сторону, что не является ошибками и не окаже значительного влияния на статистику по данному показателю.

2) По пульсу не совсем верно определились интервалы определения выбросов, тк одно значение ушло ниже нуля, что невозможно. Также нашлись выбросы превышающие пороговое значение (3), но при рассмотрении можно увидеть, что значения могут юыть связаны с высокой температурой или же высоким уровнем боли. Не считаем ошибкой, оставляем.

3) По наличию красных клеток в крови все ок, одно значение выше порога, превышает незначительно.

4) Получились слишком широкие рамки для попадания стандартных значений, в связи с чем не произошло выбросов.

5) Также было определено, что в столбце age была допущена ошибка, предположительно вместо категории 2 была проставлена категория 9, с что надо будет проработать и поменить значения 9 на 2.


## Температура

In [38]:
q1t = main_factors.rectal_temperature.quantile(0.25)
q3t = main_factors.rectal_temperature.quantile(0.75)
iqrt = q3t - q1t
lower_bound_t = q1t - (1.5 * iqrt) 
upper_bound_t = q3t + (1.5 * iqrt)

remove_outliers_t = main_factors[main_factors.rectal_temperature.between(lower_bound_t, upper_bound_t, inclusive=True)].sort_values('rectal_temperature')

print(lower_bound_t, upper_bound_t)

36.74999999999999 39.550000000000004


In [39]:
pd.concat([main_factors, remove_outliers_t]).drop_duplicates(keep=False).dropna()

surgery  age  rectal_temperature  pulse  pain  peristalsis  \
20       1.0    1                39.9   72.0   5.0          4.0   
54       2.0    1                40.3  114.0   2.0          3.0   
80       1.0    1                36.4   98.0   4.0          3.0   
91       2.0    1                40.3  114.0   2.0          3.0   
99       2.0    1                39.6  108.0   2.0          4.0   
118      1.0    1                36.5   78.0   5.0          3.0   
238      2.0    1                36.1   88.0   3.0          3.0   
251      2.0    1                36.6   42.0   1.0          4.0   
259      1.0    1                40.8   72.0   2.0          3.0   
281      2.0    1                40.0   78.0   2.0          3.0   
298      1.0    1                36.5  100.0   3.0          3.0   

     abdominal_distension  packed_cell_volume  total_protein  outcome  
20                    4.0                46.0            6.1      1.0  
54                    3.0                57.0            8.1      3.0  
80                    2.0                47.0            6.4      2.0  
91                    3.0                57.0            8.1      2.0  
99                    3.0                59.0            8.0      1.0  
118                   1.0                34.0           75.0      1.0  
238                   2.0                45.0            7.0      3.0  
251                   1.0                52.0            7.1      2.0  
259                   1.0                54.0            7.4      2.0  
281                   1.0                66.0            6.5      2.0  
298                   3.0                50.0            6.0      1.0

## Пульс

In [40]:
q1p = main_factors.pulse.quantile(0.25)
q3p = main_factors.pulse.quantile(0.75)
iqrp = q3p - q1p
lower_bound_p = q1p - (1.5 * iqrp) 
upper_bound_p = q3p + (1.5 * iqrp)

remove_outliers_p = main_factors[main_factors.pulse.between(lower_bound_p, upper_bound_p, inclusive=True)].sort_values('pulse')

print(lower_bound_p, upper_bound_p)

-12.0 148.0


In [41]:
pd.concat([main_factors, remove_outliers_p]).drop_duplicates(keep=False).dropna()

surgery  age  rectal_temperature  pulse  pain  peristalsis  \
3        1.0    9                39.1  164.0   2.0          4.0   
255      1.0    9                38.8  184.0   4.0          1.0   
275      1.0    9                38.8  150.0   5.0          3.0   

     abdominal_distension  packed_cell_volume  total_protein  outcome  
3                     4.0                48.0            7.2      2.0  
255                   3.0                33.0            3.3      2.0  
275                   2.0                50.0            6.2      2.0

## Красные клетки

In [42]:
q1c = main_factors.packed_cell_volume.quantile(0.25)
q3c = main_factors.packed_cell_volume.quantile(0.75)
iqrc = q3c - q1c
lower_bound_c = q1c - (1.5 * iqrc) 
upper_bound_c = q3c + (1.5 * iqrc)

remove_outliers_c = main_factors[main_factors.packed_cell_volume.between(lower_bound_c, upper_bound_c, inclusive=True)].sort_values('packed_cell_volume')

print(lower_bound_c, upper_bound_c)

17.0 73.0


In [43]:
pd.concat([main_factors, remove_outliers_c]).drop_duplicates(keep=False).dropna()

surgery  age  rectal_temperature  pulse  pain  peristalsis  \
144      1.0    1                37.1   84.0   2.0          4.0   

     abdominal_distension  packed_cell_volume  total_protein  outcome  
144                   4.0                75.0           81.0      3.0

## Протеин

In [44]:
q1tp = main_factors.total_protein.quantile(0.25)
q3tp = main_factors.total_protein.quantile(0.75)
iqrtp = q3tp - q1tp
lower_bound_tp = q1tp - (1.5 * iqrtp) 
upper_bound_tp = q3tp + (1.5 * iqrtp)

remove_outliers_tp = main_factors[main_factors.total_protein.between(lower_bound_tp, upper_bound_tp, inclusive=True)].sort_values('total_protein')

print(lower_bound_tp, upper_bound_tp)

-69.25 132.75


In [45]:
pd.concat([main_factors, remove_outliers_tp]).drop_duplicates(keep=False).dropna()

Empty DataFrame
Columns: [surgery, age, rectal_temperature, pulse, pain, peristalsis, abdominal_distension, packed_cell_volume, total_protein, outcome]
Index: []

In [49]:
main_factors.age.unique()

array([1, 9])

## Задание 3

Заново расчитываем количество пропусков на выбранных колонках обновленного дата фрейма.

In [67]:
import numpy as np

In [56]:
main_factors.info()

i = 1
for col in main_factors.columns:
    pct_missing = main_factors[col].isnull().mean()
    print(f'{i} {col} - {pct_missing :.1%}')
    i += 1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   age                   300 non-null    int64  
 2   rectal_temperature    240 non-null    float64
 3   pulse                 276 non-null    float64
 4   pain                  245 non-null    float64
 5   peristalsis           256 non-null    float64
 6   abdominal_distension  244 non-null    float64
 7   packed_cell_volume    271 non-null    float64
 8   total_protein         267 non-null    float64
 9   outcome               299 non-null    float64
dtypes: float64(9), int64(1)
memory usage: 23.6 KB
1 surgery - 0.3%
2 age - 0.0%
3 rectal_temperature - 20.0%
4 pulse - 8.0%
5 pain - 18.3%
6 peristalsis - 14.7%
7 abdominal_distension - 18.7%
8 packed_cell_volume - 9.7%
9 total_protein - 11.0%
10 outcome - 0.3%


Для начала нужно посмотреть, есть ли строки с большим количеством пропусков. Если у нас будет отсутствовать информация по 4-10 из 10 показателей, то вряд ли эта строка будет репрезентативной, такие строки мы удалим. Таким образом мы удаляем 16 строк.

Также я считаю, что вместо значения 2 в колонке возраст было проставлено значение 9. Предполагаю, что это просто ошибка, также если это количество лет или месяцев, то все равно эти лошади будут относиться к 2 группе (но это вряд ли, тк есть только значение 9 и 1, прочих нет).

Далее пройдемся по всем колонкам:

У нас есть одно незаполненное поле по наличию операцииотсмотрим что это за лошадь и можно ли определить это по другим строкам.

Тк температура идет с небольшим разбросом, мы можем ее заполнить средними значениями. Также можем предположить от чего зависит температура, разбить нашу совокупность на группы в соответствии с этими факторами, и заполнить средними значениями по этим группам.

По пульсу определим группы по уровню боли и заполним средними по группам значениям.



In [130]:
main_factors_nn = main_factors.dropna(thresh = 7)

In [131]:
# Изменение значений по колонке возраста
main_factors_nn.age = np.where(main_factors_nn['age'] == 9, 2, main_factors_nn['age'])

In [176]:
main_factors_nn.info()

i = 1
for col in main_factors_nn.columns:
    pct_missing = main_factors_nn[col].isnull().mean()
    print(f'{i} {col} - {pct_missing :.1%}')
    i += 1

<class 'pandas.core.frame.DataFrame'>
Int64Index: 276 entries, 0 to 299
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               276 non-null    float64
 1   age                   276 non-null    int64  
 2   rectal_temperature    233 non-null    float64
 3   pulse                 263 non-null    float64
 4   pain                  238 non-null    float64
 5   peristalsis           252 non-null    float64
 6   abdominal_distension  239 non-null    float64
 7   packed_cell_volume    276 non-null    float64
 8   total_protein         276 non-null    float64
 9   outcome               276 non-null    float64
dtypes: float64(9), int64(1)
memory usage: 33.7 KB
1 surgery - 0.0%
2 age - 0.0%
3 rectal_temperature - 15.6%
4 pulse - 4.7%
5 pain - 13.8%
6 peristalsis - 8.7%
7 abdominal_distension - 13.4%
8 packed_cell_volume - 0.0%
9 total_protein - 0.0%
10 outcome - 0.0%


In [79]:
# Ищем по какой записи не прописалось наличие операции
# Данная лошадь не репрезентативна, тк мы не значеи была ли проведена операция и какой был исход, стоит удалить

In [133]:
main_factors_nn[main_factors_nn[['surgery']].isnull().any(1)]

surgery  age  rectal_temperature  pulse  pain  peristalsis  \
132      NaN    1                38.0   48.0   4.0          3.0   

     abdominal_distension  packed_cell_volume  total_protein  outcome  
132                   1.0                43.0           73.0      NaN

In [134]:
main_factors_nn.drop([132], inplace = True)

/Applications/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:3990: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [119]:
# Предполагаем, что температура зависит от зараженности, как написано в описании данных. Но тк у нас нет данных о зараженности, будем считать что температура зависит от шока, и, соответственно, от болевых ощущений.
# Но тк болевые ощущения у нас также пока не полностью

main_factors_nn.rectal_temperature.fillna(main_factors_nn.groupby('pain').rectal_temperature.transform('mean'), inplace = True)

In [121]:
main_factors_nn[main_factors_nn[['rectal_temperature']].isnull().any(1)]

surgery  age  rectal_temperature  pulse  pain  peristalsis  \
7        1.0    1                 NaN   60.0   NaN          4.0   
68       2.0    1                 NaN   52.0   NaN          NaN   
87       2.0    1                 NaN   40.0   NaN          NaN   
96       2.0    1                 NaN   84.0   NaN          3.0   
152      1.0    1                 NaN  100.0   NaN          4.0   
274      1.0    1                 NaN   76.0   NaN          4.0   
293      1.0    1                 NaN   78.0   NaN          3.0   

     abdominal_distension  packed_cell_volume  total_protein  outcome  
7                     2.0                44.0            8.3      2.0  
68                    NaN                55.0            7.2      1.0  
87                    NaN                50.0            7.0      3.0  
96                    1.0                44.0            8.5      1.0  
152                   4.0                68.0           64.0      1.0  
274                   4.0                 NaN            NaN      3.0  
293                   NaN                43.0           62.0      3.0

In [123]:
print(main_factors_nn.groupby('surgery').rectal_temperature.mean())

surgery
1.0    38.155868
2.0    38.207699
Name: rectal_temperature, dtype: float64


In [137]:
# Предполагаем, что уровень боли зависит от пульса, что частично подтверждается данными ниже
# Будем заполнять по интервалам, предположим что пульс 30-55 относится к первой группе боли, 55-65 к 3, 65-80 к 2, 80-85 к 4 и 85 и выше к 5. 
# Также мы значем, что для всех незаполненных значений боли у нас есть значения пульса, что приведет к полному заполнению значений боли

print(main_factors_nn.groupby('pain').pulse.mean())
main_factors_nn[main_factors_nn[['pain']].isnull().any(1)]

pain
1.0    50.081081
2.0    80.859649
3.0    63.876923
4.0    84.151515
5.0    88.606061
Name: pulse, dtype: float64


surgery  age  rectal_temperature  pulse  pain  peristalsis  \
4        2.0    1                37.3  104.0   NaN          NaN   
7        1.0    1                 NaN   60.0   NaN          4.0   
19       2.0    1                39.4  110.0   NaN          3.0   
24       1.0    1                38.1   60.0   NaN          4.0   
27       1.0    1                37.8   48.0   NaN          3.0   
36       2.0    1                38.3  112.0   NaN          NaN   
41       2.0    2                39.0  150.0   NaN          NaN   
53       2.0    1                38.6   40.0   NaN          NaN   
94       2.0    1                37.5   48.0   NaN          2.0   
96       2.0    1                 NaN   84.0   NaN          3.0   
106      1.0    1                38.3   52.0   NaN          NaN   
129      1.0    1                38.4   49.0   NaN          NaN   
141      2.0    1                36.0   42.0   NaN          NaN   
152      1.0    1                 NaN  100.0   NaN          4.0   
157      1.0    1                37.8   60.0   NaN          4.0   
161      1.0    1                38.3   42.0   NaN          NaN   
165      1.0    1                39.4  120.0   NaN          3.0   
170      2.0    1                37.5   44.0   NaN          1.0   
185      1.0    1                38.5  120.0   NaN          3.0   
186      1.0    1                39.3   64.0   NaN          3.0   
188      1.0    1                38.5   60.0   NaN          1.0   
195      2.0    1                38.0   44.0   NaN          1.0   
202      1.0    1                37.5   48.0   NaN          NaN   
208      1.0    1                37.8   88.0   NaN          3.0   
213      1.0    1                37.6   45.0   NaN          2.0   
217      1.0    1                38.0   76.0   NaN          NaN   
229      1.0    2                38.5  120.0   NaN          1.0   
232      1.0    1                38.5   30.0   NaN          NaN   
234      2.0    1                37.5   48.0   NaN          2.0   
243      1.0    1                38.4   92.0   NaN          3.0   
257      1.0    1                38.7   96.0   NaN          4.0   
262      2.0    2                38.6   88.0   NaN          NaN   
268      1.0    1                38.2   42.0   NaN          3.0   
269      1.0    1                39.3   64.0   NaN          3.0   
283      1.0    1                38.2   72.0   NaN          NaN   
287      2.0    2                39.5   84.0   NaN          NaN   
293      1.0    1                 NaN   78.0   NaN          3.0   
299      1.0    1                37.2   40.0   NaN          NaN   

     abdominal_distension  packed_cell_volume  total_protein  outcome  
4                     NaN                74.0            7.4      2.0  
7                     2.0                44.0            8.3      2.0  
19                    3.0                55.0            8.7      1.0  
24                    3.0                51.0           65.0      1.0  
27                    2.0                37.0            5.5      1.0  
36                    1.0                51.0            6.0      3.0  
41                    NaN                47.0            8.5      1.0  
53                    NaN                41.0            6.4      1.0  
94                    1.0                48.0            8.6      1.0  
96                    1.0                44.0            8.5      1.0  
106                   NaN                43.0            6.1      1.0  
129                   1.0                44.0            7.6      1.0  
141                   NaN                64.0            6.8      2.0  
152                   4.0                68.0           64.0      1.0  
157                   4.0                52.0           75.0      3.0  
161                   NaN                40.0            8.5      2.0  
165                   3.0                56.0           64.0      3.0  
170                   1.0                35.0            7.2      1.0  
185                   N

In [147]:
# Заполняем значения для боли

main_factors_nn.pain = np.where((main_factors_nn.pain.isnull) & (55 > main_factors_nn.pulse > 30), 1, main_factors_nn.pain)




ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [155]:
def pain(pa, pu):
    if pa.isnull & (55 > pu >= 30):
        return 1
    elif pa.isnull & (80 > pu >= 65):
        return 2
    elif pa.isnull & (65 > pu >= 55):
        return 3
    elif pa.isnull & (85 > pu >= 80):
        return 4
    else:
        return 5
    
main_factors_nn.pain = main_factors_nn.apply(pain(main_factors_nn.pain, main_factors_nn.pulse), axis = 1)
        
    

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [ ]:
## Заполняем значение perestalsis

In [163]:
## Заполняем значение для красных клеток
print(main_factors_nn.groupby(['surgery', 'age']).packed_cell_volume.mean())
main_factors_nn.packed_cell_volume.fillna(main_factors_nn.groupby(['surgery', 'age']).packed_cell_volume.transform('mean'), inplace=True)

surgery  age
1.0      1      47.407801
         2      42.866667
2.0      1      46.143564
         2      37.333333
Name: packed_cell_volume, dtype: float64


/Applications/anaconda3/lib/python3.8/site-packages/pandas/core/generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [175]:
## Заполняем значение для протеина медианой, тк есть значительные правые хвосты для группы с исходом = 1

print(main_factors_nn.groupby(['age', 'outcome']).total_protein.median())
main_factors_nn.total_protein.fillna(main_factors_nn.groupby(['age', 'outcome']).total_protein.transform('median'), inplace=True)



age  outcome
1    1.0         7.6
     2.0         6.9
     3.0        57.5
2    1.0         6.2
     2.0         6.4
     3.0        57.0
Name: total_protein, dtype: float64
